### DISCUSSION TITLE
모바일에서 Raycast로 Hit 했을 때 GetComponent를 가져오지 못하는 원인 아시는 분 있나요?

### DISCUSSION THREAD

**Discussion by ryusiin** [naverz/zepeto-studio-kor#528] [5. Q&A]
=====

// :: NPC Data 획득
const npcData = hitInfo.collider.gameObject.GetComponent<NPCData>();

=====

![image](https://user-images.githubusercontent.com/5895396/159865250-367e4710-0b16-42ce-80d6-dd2370e34643.png)

PC에서는 해당 스크립트를 잘 가져오는데,
모바일에서는 해당 스크립트를 제대로 가져오지 못하는 현상이 있습니다.

NPC가 움직이는 것을 보면 NPC에게 일단 해당 스크립트가 부착되어 있는 것 같습니다.

===== 전체 코드 =====

private * ShootRay() {
        while (true) {
            // :: 스타팅 포지션
            const startingPosition = new UnityEngine.Vector3(
                this.gameObject.transform.position.x,
                this.gameObject.transform.position.y + 0.5,
                this.gameObject.transform.position.z
            );

            // :: 레이 제작
            let ray: UnityEngine.Ray = new UnityEngine.Ray(
                startingPosition, this.gameObject.transform.forward);

            // RaycatHit 참조변수를 파라메터로 넘기기 위해, $ref 연산자로 wrapping 하는 부분에 유의해주세요.
            // (매개변수 한정자 out이 필요한 상황이기때문에) 객체 생성 없이 참조만 선언해줍니다.            
            let ref = $ref<UnityEngine.RaycastHit>();

            // :: 레이어 마스크
            const layerMask = (1 << UnityEngine.LayerMask.NameToLayer("NPC"))
                + (1 << UnityEngine.LayerMask.NameToLayer("Player"));

            if (UnityEngine.Physics.Raycast(ray, ref, 1000, layerMask)) {
                // RaycatHit 데이터를 확인하기 위해, $unref 연산자로 다시 unwrapping 하는 부분에 유의해주세요.
                let hitInfo = $unref(ref);

                // :: NPC Data 획득
                const npcData = hitInfo.collider.gameObject.GetComponent<NPCData>();
                // :: 충돌 확인
                PastaRuler.oInstance.UIManager.CheckCollider(true);
                if (npcData != null) {

                    if (this.iCurTarget_NPC != null) {
                        this.iCurTarget_NPC.HideTargetCube();
                    }

                    this.iCurTarget_NPC = npcData;
                    this.iCurTarget_NPC.ShowTargetCube();
                } else {
                    hitInfo.collider.gameObject.transform.position =
                        new UnityEngine.Vector3(
                            hitInfo.collider.gameObject.transform.position.x,
                            hitInfo.collider.gameObject.transform.position.y + 1,
                            hitInfo.collider.gameObject.transform.position.z
                        );
                }
            } else {
                // :: 충돌 확인
                PastaRuler.oInstance.UIManager.CheckCollider(false);

                if (this.iCurTarget_NPC != null) {
                    this.iCurTarget_NPC.HideTargetCube();
                    this.iCurTarget_NPC = null;
                }

                UnityEngine.Debug.DrawRay(
                    startingPosition, this.gameObject.transform.forward,
                    UnityEngine.Color.red, 0.1);
            }

            yield new UnityEngine.WaitForSeconds(0.1);
        }
    }

==========

**Comment by Zepeto-TechEvangelist** [comment]
안녕하세요, 런타임에 추가된 커스텀 클래스 컴포넌트일 경우 모바일 테스트시에 
GetComponent가 안되는 문제가 보고되어 파악 중에 있습니다.
감사합니다.

**Reply by ryusiin** [reply]
그렇군요...:>
감사합니다, 다른 방법을 찾아봐야겠군요

**Reply by hawai0622** [reply]
GetComponent가 null인 동일한 문제가 발생하고 있습니다.
커스텀 클래스에서 GetComponent가 안되는 문제는 수정 진행중인가요?

**Reply by Chordcode002** [reply]
이 문제 해결되었나요? 지금도 똑같은 문제가 발생하네요...

**Reply by Zepeto-TechEvangelist** [reply]
안녕하세요, 수정 진행 중에 있으며 수정 완료될 경우 패치 노트에 공개 예정입니다.
다만 적용 시기는 현재로서 공유드리기 어려운 점 양해 부탁 드립니다.

**Reply by Zepeto-TechEvangelist** [reply]
안녕하세요, 해당 이슈 월드 버전 1.5.0에 수정 작업 반영될 예정입니다.
감사합니다.

**Comment by cheolhui-lee** [comment]
안녕하세요.
일단 PC에서는 되는데 모바일에서는 안되는 상황과 첨부해주신 코드를 종합해보면 Layer의 이슈가 아닐까 생각이 됩니다.
의도대로 동작하지 않은 Layer 설정으로 인해 잘못된 오브젝트가 hit 될 수 있어 보입니다.
유니티 에디터에서 직접 추가하신 것으로 보이는 "NPC", "Player" 등의 Layer는 당연히 제페토 앱에는 탑재가 되어 있지 않기 때문에 저희 SDK 내부적으로 매칭을 시켜주는 로직이 있는데요.
아마 관련해서 LayerMask.NameToLayer("Player") 같은 로직이 비정상 동작하는지 저희가 확인을 해봐야 할 것 같습니다..
그리고 확인에 도움이 될 수 있도록 아래 이미지와 같이 현재 등록하신 전체 레이어 정보를 알려주시면 저희가 문제 해결을 하는데 도움이 될 것 같습니다.
![image](https://user-images.githubusercontent.com/3862666/160794633-e24a0a22-389a-47dd-b95a-25f187416bfc.png)
